examples/dxc/src/main.rs

```diff
@@ -0,0 +1,226 @@
+    assert_eq!(cp, DXC_CP_UTF8);
+    unsafe {
+        let slice = std::slice::from_raw_parts(
+            blob.GetBufferPointer() as *const u8,
+            blob.GetBufferSize() - 1,
+        );
+        std::str::from_utf8_unchecked(slice)
+    }
+}
+
+fn create_blob(library: &IDxcLibrary, data: &str) -> windows::Result<IDxcBlobEncoding> {
+    let mut blob = None;
+    unsafe {
+        library.CreateBlobWithEncodingFromPinned(
+            data.as_ptr() as *const _,
+            data.len() as u32,
+            DXC_CP_UTF8,
+            &mut blob,
+        )
+    }
+    .and_some(blob)
+}
+
+#[allow(non_snake_case)]
+fn main() -> windows::Result<()> {
+    let lib = unsafe { Library::new(dxcompiler_lib_name()) }.unwrap();
+    let create: Symbol<DxcCreateInstanceProc> = unsafe { lib.get(b"DxcCreateInstance\0") }.unwrap();
+    dbg!(&create);
+
+    // let mut compiler: Option<IDxcCompiler2> = None;
+    // let compiler = unsafe { create(&CLSID_DxcCompiler, &IDxcCompiler2::IID, compiler.set_abi()) }
+    //     .and_some(compiler)
+    //     .unwrap();
+    let compiler: IDxcCompiler2 = unsafe { DxcCreateInstanceProc(&create, &CLSID_DxcCompiler) }?;
+    let library: IDxcLibrary = unsafe { DxcCreateInstanceProc(&create, &CLSID_DxcLibrary) }?;
+
+    dbg!(&compiler, &library);
+
+    let main_shader = "#include \"copy.hlsl\"";
+    let shader_blob = create_blob(&library, main_shader)?;
+
+    unsafe extern "system" fn LoadSource(
+        this: RawPtr,
+        pfilename: PWSTR,
+        ppincludesource: *mut RawPtr,
+    ) -> HRESULT {
+        let this = &mut *(this as *mut IncludeHandlerData);
+        dbg!(&this, pfilename, ppincludesource);
+
+        if pfilename != "./foo/bar/copy.hlsl" {
+            return HRESULT::from_win32(ERROR_FILE_NOT_FOUND.0);
+        }
+        let copy_shader = include_str!("copy.hlsl");
+        let shader_blob = create_blob(&this.library, copy_shader).unwrap();
+        *ppincludesource = shader_blob.abi();
+        this.alive_shaders.push(shader_blob);
+        HRESULT(0)
+    }
+
+    unsafe extern "system" fn QueryInterface(
+        _ptr: RawPtr,
+        _iid: &Guid,
+        _interface: *mut RawPtr,
+    ) -> HRESULT {
+        todo!()
+    }
+
+    unsafe extern "system" fn AddRef(this: ::windows::RawPtr) -> u32 {
+        let this = &mut *(this as *mut IncludeHandlerData);
+        dbg!(this.refs.add_ref())
+    }
+
+    unsafe extern "system" fn Release(this: ::windows::RawPtr) -> u32 {
+        let this = &mut *(this as *mut IncludeHandlerData);
+        dbg!(this.refs.release())
+    }
+
+    #[cfg(not(windows))]
+    unsafe extern "system" fn dtor(_ptr: ::windows::RawPtr) {
+        todo!()
+    }
+
+    let include_handler = IDxcIncludeHandler_abi(
+        // IUnknown
+        QueryInterface,
+        AddRef,
+        Release,
+        // Virtual destructors breaking the vtable, hack for !windows DXC
+        #[cfg(not(windows))]
+        dtor,
+        #[cfg(not(windows))]
+        dtor,
```

Comment:
This should probably be removed - I'll have to take this up with DXC but it's a tricky one. For more details: https://github.com/Traverse-Research/hassle-rs/pull/5/commits/c37cb037be072cac43f2706b89f1388b9deee19a#diff-8202c69531938d4c80a8f8a9190be8b98a026d2bd0364b97a0c7a32e8ae21ac6

Reply:
Thanks, that'd be nice. Non-Windows support is fairly recent and it looks mostly used for the `./dxc` binary, not the library. I've already submitted some fixes over time to make the library workable, and this seems to be the final step to make it truly cross-compatible. I [mentioned this issue](https://github.com/microsoft/DirectXShaderCompiler/issues/2680#issuecomment-667676936) a while ago but didn't get to report a standalone issue yet nor dig and fix it myself.

> Looks like IUnknown explicitly has a virtual destructor on non-Windows.

Correct: https://github.com/microsoft/DirectXShaderCompiler/blob/aa4f5176ebba010a860a214d73f33bdc5b5042ce/include/dxc/Support/WinAdapter.h#L627

I already tried removing it but that results in tons of compiler warnings, though it _should_ work since on Windows these objects aren't cleaned up through a virtual DTOR either... I'll revisit this again on both GCC and Clang, and see what happens. My C++ knowledge isn't up there so I don't know if we can simply disable the warning without breaking anything. Note that `Release` calls `delete` on itself which would call the destructor and deallocate: https://github.com/microsoft/DirectXShaderCompiler/blob/aa4f5176ebba010a860a214d73f33bdc5b5042ce/lib/DxcSupport/WinAdapter.cpp#L24. Aside, the vtable clearly shows a destructor specific to `DxcLibrary` even though it has no members that need destructing - that already happens the release function. I vaguely recall reading something about passing the pointer _and size_ to `delete` under the hood, that might be why `DxcLibrary` has its own `deleting destructor` in place.

Reply:
Finally found some time to report this at 2AM: https://github.com/microsoft/DirectXShaderCompiler/issues/3783

Reply:
@damyanp For the record (in case you're still asking around) there's now a PR open: https://github.com/microsoft/DirectXShaderCompiler/pull/3793

Removing the destructor is totally fine, and even MSVC complains about missing virtual dtors on the "official" `IUnknown` from the Windows headers, if enough warnings are enabled: https://godbolt.org/z/hKPT6ThEf